Sources/Persistence/DiskMemoryPersistenceStack.swift

```diff
@@ -1,458 +1,429 @@
-        $0.name = "com.mindera.alicerce.persistence.diskmem.write"
-        $0.maxConcurrentOperationCount = 1
-        return $0
-    }(OperationQueue())
+        private let configuration: Configuration
 
-    private var usedDiskSize: UInt64 = 0 // Size in bytes
+        private var readQueue: DispatchQueue? { return configuration.readQueue }
 
-    // needs to be atomic since `NSCache` operations are not always made on the same queue/thread, i.e. caller
-    private let usedMemorySize = Atomic<UInt64>(0) // Size in bytes
+        private let writeOperationQueue: OperationQueue = {
+            $0.name = "com.mindera.alicerce.persistence.diskmem.write.operation-queue"
+            $0.maxConcurrentOperationCount = 1
+            return $0
+        }(OperationQueue())
 
-    public init(configuration: Configuration) {
-        self.configuration = configuration
+        private var usedDiskSize: UInt64 = 0 // Size in bytes
 
-        super.init()
+        // needs to be atomic since `NSCache` operations are not always made on the same queue/thread, i.e. caller
+        private let usedMemorySize = Atomic<UInt64>(0) // Size in bytes
 
-        cache.totalCostLimit = Int(configuration.memLimit)
-        cache.delegate = self
+        public init(configuration: Configuration) throws {
+            self.configuration = configuration
 
-        readOperationQueue.qualityOfService = configuration.qos.read
-        writeOperationQueue.qualityOfService = configuration.qos.write
+            super.init()
 
-        diskCacheEnabled = hasDiskCacheDirectory() || createDiskCacheDirectory()
+            cache.totalCostLimit = Int(configuration.memLimit)
+            cache.delegate = self
 
-        calculateUsedDiskSize()
-    }
+            writeOperationQueue.underlyingQueue = configuration.writeQueue
 
-    // MARK: - Public Methods
+            if hasDiskCacheDirectory() == false {
+                try createDiskCacheDirectory()
+            }
 
-    public func object(for key: Persistence.Key, completion: @escaping PersistenceCompletionClosure<Data>) {
-        if let data = cachedData(for: key) {
-            return completion { data }
+            calculateUsedDiskSize()
         }
 
-        // TODO: Check for object on disk and cache it if don't available
+        deinit {
+            writeOperationQueue.waitUntilAllOperationsAreFinished()
+        }
 
-        diskData(for: key) { [weak self] inner in
-            do {
-                let data = try inner()
+        // MARK: - Public Methods
 
-                self?.setCachedData(data, for: key)
+        public func object(for key: Key, completion: @escaping ReadCompletionClosure) {
+            if let data = cachedData(for: key) {
+                return completion(.success(data))
 
-                completion { data }
-            } catch {
-                completion { throw error }
+                // TODO: Check for object on disk and cache it if not present
```

Comment:
Isn't this already being done with `setCachedData` below?

Reply:
Initially I also had to think a bit to understand what it was (and also thought it was already done), so I improved the comment a bit to make it clearer what the `TODO` really is 😉👍